### ISSUE TITLE
Problems with window customization's rounded corners and shadows

### ISSUE THREAD

**Issue by sessionboy** [tauri-apps/tauri#9287]
### Describe the bug

Because tauri does not support the extension of native windows, I have to manually implement custom windows.  I tried to simulate macos' window implementation and expand on it, such as menus, tabs, and titles.
I implemented some basic functions through the [tauri-controls](https://github.com/agmmnn/tauri-controls) library, but there are still some issues.
### 1. Incomplete implementation of rounded corners
To support rounded corners, I made some settings:
```json
"windows": [
      {
        "title": "myapp",
        "shadow": true,
        "width": 1000,
        "height": 700,
        "fullscreen": false,
        "center": true,
        "transparent": true,
        "decorations": false,
        "resizable": true
      }
 ],
```
and then (styles.css):
```css
:root {
    border-radius: 12px;
    ...
}
body{
    background: #fff;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border: none;
    margin: 0;
    padding: 0;
    overflow: hidden;
}
```
The final effect:
![11111](https://github.com/tauri-apps/tauri/assets/17827693/14f36526-c0a3-4ede-90c2-af6a364d5921)

Obviously, the corners on the top left and top right are not rounded.  The background seems to be transparent, but it still has a border that affects the implementation of rounded corners. How can I fix it?

### 2. Shadow enhancement
I want to make some modifications to the shadow to enhance it. How can I do that?
![2222](https://github.com/tauri-apps/tauri/assets/17827693/70b5b54f-0b4d-4aec-9ef3-0dcdb1b374da)

About tauri v2, I hope the official can provide a complete case of customizing windows, including simulating native windows, as well as customizing menus, tabs, titles, and other features.

### Full `tauri info` output

```text
nothing
```

### Additional context

system: macOS monterey
version: 12.6
tauri version:
```toml
tauri = { version = "2.0.0-beta.13", features = ["macos-private-api"] }
tauri-plugin-os = "2.0.0-beta.3"
tauri-plugin-shell = "2.0.0-beta.3"
```
```json
{
   "@tauri-apps/api": "v2.0.0-beta.6",
    "@tauri-apps/plugin-os": "2.0.0-beta.2",
    "@tauri-apps/cli": "2.0.0-beta.11"
}
```

**Comment by benjamingorman** [comment]
Also have the same problem. Interested in the solution to this!

**Comment by KJohnson82** [comment]
I've been fighting this issue for a day and a half, no joy so far, hopefully someone here can help

**Comment by nianyi778** [comment]
我i也很感兴趣

**Comment by kodgurkini** [comment]
For anyone still searching for a solution to this, set "shadow": false

**Comment by KJohnson82** [comment]
> For anyone still searching for a solution to this, set "shadow": false

Can you give a little more information, please?

**Comment by kodgurkini** [comment]
Absolutely, in the tauri.conf.json, or as you create a new window, you can set window properties like:

"windows": [
      {
        "height": 300,
        "width": 240,
        "title": "Title of your app",
        "decorations": false, 
        "transparent": true,
        "shadow": false
      }
    ],

Adding the "shadow": false here removes the border-looking-shadow around the app. Don't forget to set body bg to transparent etc. Hope that helps!

**Comment by KJohnson82** [comment]
> Absolutely, in the tauri.conf.json, or as you create a new window, you can set window properties like:
> 
> "windows": [ { "height": 300, "width": 240, "title": "Title of your app", "decorations": false, "transparent": true, "shadow": false } ],
> 
> Adding the "shadow": false here removes the border-looking-shadow around the app. Don't forget to set body bg to transparent etc. Hope that helps!

Ah okay, I think that is a feature of 2.0, my app will need to stay on 1.0 for a little while as I need to support windows 7, I will check on the CSS and tailwind to make sure shadows are not an issue and that might fix the problem. Thanks for the help!

**Comment by tzfun** [comment]
I encountered the same problem, which only occurs on Windows platform and after using set_shadow. My tauri version is v1. After removing the dependency on set_shadow, it is fine. It seems that it is caused by setting the window shadow. Unfortunately, I can only give up the rounded corners and references. I hope the tauri team can fix this problem.

**Comment by xiaoyan2015** [comment]
![未标题-3](https://github.com/user-attachments/assets/d402a34a-55af-43be-bf5b-9629a60e4cc8)
This is how I implemented it。

**Comment by JiuRanYa** [comment]
I got the same problem. The solution above seems not work.

**Comment by Schachte** [comment]
I have this working on Mac.. maybe give this a shot @JiuRanYa 

_Example:_
![image](https://github.com/user-attachments/assets/38259b72-5035-42ac-996d-ca6d960ccf05)

_tauri.conf.json_
```
  "app": {
    "macOSPrivateApi": true,
    "windows": [
      {
        "label": "swip-swap",
        "title": "swip-swap",
        "width": 800,
        "height": 50,
        "resizable": true,
        "fullscreen": false,
        "decorations": false,
        "transparent": true,
        "shadow": false
      }
    ],
    "security": {
      "csp": null,
      "capabilities": [
        {
          "identifier": "draggable-window",
          "windows": ["*"],
          "permissions": ["core:window:allow-start-dragging"]
        }
      ]
    }
  },
```

_index.css_
```
body {
  margin: 0;
  display: flex;
  place-items: center;
  border-radius: 10px;
  width: 100%;
  height: 100%;
}

.drag-wrapper {
  display: flex;
  width: 100%;
}

#root {
  padding: 1rem;
  border-radius: 50px;
  height: 100vh;
  background-color: #f3f3f3;
  border-radius: 5px;
  width: 100%;
  height: 100%;
}
```

**Comment by JiuRanYa** [comment]
@Schachte First at all, so Appreciated for your reply.

But I'm using tauri v2.0.0, I tried your code, but seems not work. I got this:
 
![image](https://github.com/user-attachments/assets/306b72b6-e407-43c8-9cc5-0dbe22f117e6)

**Comment by ghost** [comment]
I believe you have to clip overflow

**Comment by AssafShabtay** [comment]
> @Schachte First at all, so Appreciated for your reply.
> 
> But I'm using tauri v2.0.0, I tried your code, but seems not work. I got this:
> 
> ![image](https://private-user-images.githubusercontent.com/58846658/380613449-306b72b6-e407-43c8-9cc5-0dbe22f117e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NjAxNjIsIm5iZiI6MTczMzc1OTg2MiwicGF0aCI6Ii81ODg0NjY1OC8zODA2MTM0NDktMzA2YjcyYjYtZTQwNy00M2M4LTljYzUtMGRiZTIyZjExN2U2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDE1NTc0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1Y2I0MTlkMjMzNTEyZGMyOWVkN2U3ZWYyYzE0NDE5ZDRlMjYzY2VlMjRmN2ZhMWYyODU2MTgyMTA0YWFiNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-7N-onRXq-ZzC009eNZFrxJS6FtPmo-miT6xVNOIBi8)

Did you get it to work?

**Comment by Iliasnolsson** [comment]
Im also having issues with making the window transparent. Wanted to create rounded borders by first setting window as transparent. But it does not seem to work. Have tried everything mentioned in this thread.

Im using:
- Tauri v2.2.0
- macOS Sequoia 15.1

**Comment by zhanglun** [comment]
```json
"app": {
    "macOSPrivateApi": true,
    "windows": [
      {
        "fullscreen": false,
        "height": 800,
        "resizable": false,
        "title": "Pavo",
        "width": 380,
        "visible": false,
        "decorations": false,
        "transparent": true
      }
    ],
    "withGlobalTauri": true,
    "security": {
      "csp": null
    }
  }
```

the key is `macOSPrivateApi: true`

<img width="392" alt="image" src="https://github.com/user-attachments/assets/f5d66695-17a4-4f8b-837d-6b4b0f20f9d6" />

**Comment by DragorWW** [comment]
> "app": {
>     "macOSPrivateApi": true,
> }
> the key is `macOSPrivateApi: true`

@zhanglun thanks, this work for me in v2

**Comment by Schachte** [comment]
Just to note.. I'd be weary relying on these MacOS private APIs. Have used these in the past for some graphics things and they change frequently without apologies. They are unstable and undocumented in many cases. Also, I don't believe you can deploy things to the Mac App store that leverage private MacOS APIs due to their instability. 

@DragorWW @zhanglun

**Comment by zhanglun** [comment]
> Just to note.. I'd be weary relying on these MacOS private APIs. Have used these in the past for some graphics things and they change frequently without apologies. They are unstable and undocumented in many cases. Also, I don't believe you can deploy things to the Mac App store that leverage private MacOS APIs due to their instability.
> 
> [@DragorWW](https://github.com/DragorWW) [@zhanglun](https://github.com/zhanglun)

You're absolutely right. However, since my app isn't intended for the App Store, I've decided to go ahead and use them.

**Comment by niklauslee** [comment]
It seems doesn't work in MacOS Tahoe. Anyone who did it in MacOS Tahoe?

**Comment by mayurjadhav2002** [comment]
> "app": {
>     "macOSPrivateApi": true,
>     "windows": [
>       {
>         "fullscreen": false,
>         "height": 800,
>         "resizable": false,
>         "title": "Pavo",
>         "width": 380,
>         "visible": false,
>         "decorations": false,
>         "transparent": true
>       }
>     ],
>     "withGlobalTauri": true,
>     "security": {
>       "csp": null
>     }
>   }
> the key is `macOSPrivateApi: true`
> 
> <img alt="image" width="392" src="https://private-user-images.githubusercontent.com/3086358/399440430-f5d66695-17a4-4f8b-837d-6b4b0f20f9d6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA0NTExNTgsIm5iZiI6MTc2MDQ1MDg1OCwicGF0aCI6Ii8zMDg2MzU4LzM5OTQ0MDQzMC1mNWQ2NjY5NS0xN2E0LTRmOGItODM3ZC02YjRiMGYyMGY5ZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMTRUMTQwNzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWI3ZmIwYmQyMWFlNzk5MzhjZDUzMjBkZDA4Y2EwOGMxYTc2MzFjYWFlZThjZjgwZmM3M2MyNzZhM2I3MjcxMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Hk8wxDMGJacn07BJSgOMou-VP414I50S9CCc6VUDQV8">

Thanks for the solution, it worked on Mac system. Let me see if this works on Windows as well

**Comment by gajendraxdev** [comment]
Even if you round your background :root and body, your top header is still a square box that sits on top of those curves and covers them up.

The Fix:
Put a wrapper div around your entire app.
Give that wrapper `overflow: hidden` and your `border-radius`.

Note: If you use Modals, make sure they are portaled inside that wrapper, not to the document body, or they will still look square!

Here's how i doing it on my app 

`tauri.config.json`
```JSON
"app": {
    "macOSPrivateApi": true,
    "windows": [
      {
        "title": "Zync",
        "width": 1000,
        "height": 700,
        "decorations": false,
        "transparent": true,
        "shadow": false,
        "dragDropEnabled": true,
        "visible": false
      }
    ],
    "security": {
      "csp": null,
      "assetProtocol": {
        "enable": true,
        "scope": [
          "**"
        ]
      }
    }
  }
```
`MainLayout.tsx`
```tsx
// Main Layout
<div className="rounded-xl overflow-hidden bg-app-bg h-screen border border-app-border">
   <Sidebar />
   <TitleBar />
   <AppContent />
   
   {/* Internal Portal Root (Crucial for Modals) */}
   <div id="modal-portal-root" className="absolute inset-0 pointer-events-none" />
</div>
```

index.css
```css
    html,
    body,
    #root {
      width: 100%;
      height: 100%;
      margin: 0;
      /* The body must be transparent for the rounded container's 
         corners to actually show the desktop behind them */
      background: transparent !important;
    }

    body {
      overflow: hidden;
    }

```

<img width="1019" height="723" alt="Image" src="https://github.com/user-attachments/assets/a5f93d97-96ba-426f-abb1-08a3710abb32" />

<img width="1017" height="719" alt="Image" src="https://github.com/user-attachments/assets/b3621434-4af8-4f31-b4fa-e3feb76914d0" />
